src/model/UserDataSource.ts

```diff
@@ -0,0 +1,200 @@
+import { MongoDataSource } from 'apollo-datasource-mongodb'
+import { MUUID } from 'uuid-mongodb'
+import differenceInDays from 'date-fns/differenceInDays/index.js'
+
+import { getUserModel } from '../db/index.js'
+import { User, UpdateProfileGQLInput, UsernameInfo, GetUsernameReturn } from '../db/UserTypes.js'
+import { trimToNull } from '../utils/sanitize.js'
+
+const USERNAME_UPDATE_WAITING_IN_DAYS = 14
+
+interface UsernameQueryReturnType {
+  username?: string
+  createdAt?: string
+  userUuid: string
+}
+export default class UserDataSource extends MongoDataSource<User> {
+  userModel = getUserModel()
+
+  /**
+   * Update user profile.  Create a new user object if not defined.
+   * @param userUuid userUuid to update/add
+   * @param input profile params
+   * @returns true if successful
+   */
+  async createOrUpdateUserProfile (updater: MUUID, input: UpdateProfileGQLInput): Promise<boolean> {
```

Comment:
Is the idea that this would be called at signup time, alongside the current call to Auth0? And is it accurate that the long-term vision is to slowly move more and more data that is stored in the Auth0 object over into this model?

Reply:
Correct.  I'll remove the auto-generated username in Auth0 login action.  In the new flow, new users will be directed to a new username creation screen (frontend issue https://github.com/OpenBeta/open-tacos/issues/830). New users can't skip/bypass this screen because if a user is logged in, the frontend will always call `getUsername()` to see if the usernameInfo object has set.